### PR TITLE
Add tree-sitter based function, class navigation

### PR DESCRIPTION
**Pull request by sudormrfbin**
Adds support for navigating between functions, classes, etc, using tree-sitter and textobjects. Requires that a `*.around` capture is defined for a filetype in it's correspoding `textobjects.scm` file. So for goto next function to work in Rust, `function.around` textobject has to be defined in Rust's `textobjects.scm`.

helix-tree-sitter-navigation

<details>
<summary>Added keybinds (click to expand)</summary>

| Keybind | Description|
| ---| --- |
| `]f`     | Go to next function |
| `[f`     | Go to previous function |
| `]c`     | Go to next class |
| `[c`     | Go to previous class |
| `]p`     | Go to next parameter |
| `[p`     | Go to previous parameter |

</details>

#### Open questions

- Should we fall back to using the inside textobject if around is not defined ? Currently `parameter.around` is not defined in any query file, so `]p` is not very useful.

### CODE REVIEW

**Comment by EpocSquadron** [comment]
I was thinking `parameter.around`, if it were added should either select the full parameter list, or the parameter plus the delimiter (so `fn([one, ]two)`, where the square brackets represent the selection). Either way I think it would be more useful for this keybind to use `.inside` for parameters specifically.

**Comment by EpocSquadron** [comment]
How would we handle for comment navigation when I add that? Should that use around, or be more granular?

What if we have the user a choice by doing something like:

`]f` - next function body (inside)
`]F` - next complete function (around)

**Comment by sudormrfbin** [comment]
> the parameter plus the delimiter (so `fn([one, ]two)`, where the square brackets represent the selection

This was what I initially wanted to do, but I ran into [some problems](https://github.com/helix-editor/helix/pull/728/commits/5ea3f6fd2c6ad032fae221bcf78ac321eeed37b9). We would probably need something similar to neovim's custom `#make-range!` predicate to handle this properly.

> How would we handle for comment navigation when I add that? Should that use around, or be more granular?

I think around would work fine for it too, since then we can navigate by comment blocks, which seems intuitive. If we use inside for comment navigation each single line comment would be a separate node and there `j` and `k` would supersede it.

> What if we gave the user a choice by doing something like:
> 
> `]f` - next function body (inside) 
> `]F` - next complete function (around)

That IMO adds a bit too much verbosity that doesn't seem very useful since you can always use `mif` and `maf` after getting to the function.

**Comment by EpocSquadron** [comment]
@sudormrfbin ok, agreed on all counts. What about introducing a new subcapture meant specifically for this? Something like `@function.movement` that could be used alongside to get around having to fall back to `.inside` eg:

```
(anonymous_function_creation_expression
  body: (_) @function.inside) @function.around @function.movement
  
(formal_parameters
  [
    (simple_parameter)
    (variadic_parameter)
    (property_promotion_parameter)
  ] @parameter.inside @parameter.movement)

```

I just tested this out in PHP with tree-sitter-cli and it did indeed allow multiple captures on the same nodes (so @function.around and @function.movement were both captured and both corresponded to the same items):

```
  pattern: 8
    capture: 5 - parameter.inside, start: (158, 37), end: (158, 51), text: `$block = false`
    capture: 6 - parameter.movement, start: (158, 37), end: (158, 51), text: `$block = false`
  pattern: 5
    capture: function.around, start: (158, 4), end: (167, 5)
    capture: function.movement, start: (158, 4), end: (167, 5)
    capture: function.inside, start: (159, 4), end: (167, 5)
```

**Comment by EpocSquadron** [comment]
I'm also having trouble getting the parameter movements to work in PHP and Rust files after checking out this PR and trying it out. Functions and classes work beautifully.

**Comment by sudormrfbin** [comment]
Added an `object.movement` capture that can be used to override `object.around` and `object.inside`, and the precedence order is now `movement`, `around`, `inside`. I haven't added the movement capture to any file yet (with regard to parameter captures), since it works fine for now.

File: book/src/usage.md [hunk 1]
```diff
@@ -70,7 +70,26 @@ Currently supported: `word`, `surround`, `function`, `class`, `parameter`.
...
+possible by leveraging tree-sitter and textobjects queries. For
+example to move to the next function use `]f`, to move to previous
+class use `[c`, and so on.
+
+![tree-sitter-nav-demo][tree-sitter-nav-demo]
+
+See the [unimpaired][unimpaired-keybinds] section of the keybind
+documentation for the full reference.
+
+> NOTE: This feature is dependant on tree-sitter based textobjects
```
Comment (archseer) [inline book/src/usage.md:90]:
```suggestion
> NOTE: This feature is dependent on tree-sitter based textobjects
```

File: book/src/usage.md [hunk 2]
```diff
@@ -70,7 +70,26 @@ Currently supported: `word`, `surround`, `function`, `class`, `parameter`.
 | `c`                    | Class                    |
 | `p`                    | Parameter                |
 
-Note: `f`, `c`, etc need a tree-sitter grammar active for the current
+> NOTE: `f`, `c`, etc need a tree-sitter grammar active for the current
 document and a special tree-sitter query file to work properly. [Only
-some grammars](https://github.com/search?q=repo%3Ahelix-editor%2Fhelix+filename%3Atextobjects.scm&type=Code&ref=advsearch&l=&l=)
-currently have the query file implemented. Contributions are welcome !
+some grammars][lang-support] currently have the query file implemented.
+Contributions are welcome !
```
Comment (archseer) [inline book/src/usage.md:76]:
```suggestion
Contributions are welcome!
```

**Review by archseer — APPROVED** [review]
Looks good, just needs a rebase 👍🏻

**Pull request merged**